### DISCUSSION TITLE
Type eleminiation between `Key`, `PublicKey`, `PrivateKey`, and `Subkey`

### DISCUSSION THREAD

**Discussion by alichry** [openpgpjs/openpgpjs#1344] [Q&A] [answered]
Hey ✌️

I am facing an issue when it comes to type elimination, mainly to distinguish between a `Key` and `Subkey` instance. In earlier versions, `x instanceof openpgp.Key` served well. Currently, `Key` is no longer exported from `openpgp`. Therefore, we have to rely on using `x instanceof openpgp.PrivateKey || x instanceof openpgp.PublicKey` to check if `x` a primary key. However, `openpgp.readKeys` returns `openpgp.Key[]` as a type and not `Array<openpgp.PublicKey | openpgp.PrivateKey>`. 

```js
export const isPrivateKey = (key: openpgp.Key | openpgp.Subkey) => {
    if (key instanceof openpgp.PrivateKey || key instanceof openpgp.PublicKey) {
        return key.isPrivate();
    }
    return key.mainKey.isPrivate();
}
```

The above wouldn't work since we did not eliminate `Key`, only checked for specific derived classes of `Key`. If we replace the `key` type with `openpgp.PrivateKey | openpgp.PublicKey | openpgp.Subkey`, it would work well. If I want to change the types of my internal wrappers, I would have to downcast them after calling `readKeys`. 

Are there any other approaches that would be appropriate here? Do you suggest that we tweak `openpgp.readKeys`'s TS def and the like to return `Array<openpgp.PublicKey | openpgp.PrivateKey>` or `type Key = openpgp.PublicKey | openpgp.PrivateKey` ?

**Comment by larabr** [comment, marked as answer]
Hey @alichry ,
Wouldn't checking for `Subkey` explicitly work in your use case? e.g.
```js
...
    if (key instanceof Subkey) {
      return key.mainKey.isPrivate();
    }
    return key.isPrivate();
}
```
Also, `readKeys` returns `PublicKey[]`, so downcasting is only needed if you need a `PrivateKey` (whether we should keep exporting `Key` as a type is a valid question though)

**Reply by alichry** [reply]
Oh yeah, that works well ! What do you think of changing the return type of `readKeys` to `Array<openpgp.PublicKey | openpgp.PrivateKey` since that's the current implementation AFAIK and `PrivateKey` is a subclass of `PublicKey` which is a subclass of `openpgp.Key` ?

**Reply by larabr** [reply]
From the type system point of view that's redundant, since you'd still have to cast the result (if you need it, you can already do `const parsedKey: PrivateKey|PublicKey = await readKey(...);`). 

In #1347 we might switch to return a `Key` in the readKey(s) functions, to make it a bit clearer that you can downcast that to a `PrivateKey`. We are also improving support for runtime checks of the `PrivateKey` type (see the snippet in the description).

**Reply by alichry** [reply]
Thanks a lot 🤩 !!

**Reply by alichry** [reply]
Hey @larabr,

`Subkey` is not exported in [index.js](https://github.com/openpgpjs/openpgpjs/blob/e450582888f2414af3d5748845c67c036be1f849/src/index.js), hence `x instanceof openpgp.Subkey` would not work in this case. Any suggestions? I will implement a workaround for this locally.

```
Welcome to Node.js v16.2.0.
Type ".help" for more information.
> const openpgp = require('./dist/node/openpgp.min')
undefined
> openpgp.Subkey
undefined
> openpgp.SubKey
undefined
> openpgp.Key
undefined
> openpgp.PrivateKey
[class ic extends rc]
> openpgp.PublicKey
[class rc extends ec]
```

**Reply by larabr** [reply]
Hey @alichry , we can export Subkey. Feel free to open a PR, otherwise I can also do it soon :)

**Reply by alichry** [reply]
Say lesss, on it =)
